src/tasks/worker.ts

```diff
@@ -0,0 +1,263 @@
+			let error: string | undefined;
+			try {
+				await sbvrUtils.db.transaction(async (handlerTx) => {
+					const results = await handler.fn({
+						api: new PinejsClient({
+							passthrough: {
+								tx: handlerTx,
+							},
+						}),
+						params: task.is_executed_with__parameter_set ?? {},
+						tx: handlerTx,
+					});
+					status = results.status;
+					error = results.error;
+					if (results.status !== 'succeeded' && !handlerTx.isClosed()) {
+						await handlerTx.rollback();
+					}
+				});
+			} catch (err) {
+				// Ignore closed/rollback errors
+				if (!(err instanceof TransactionClosedError)) {
+					throw err;
+				}
+			} finally {
+				// Update task with results
+				await this.finalize(tx, task, startedOnTime, status, error);
+			}
+		} catch (err) {
+			// This shouldn't happen, but if it does we want to log and kill the process
+			console.error('Task execution failed:', err);
+			process.exit(1);
+		} finally {
+			this.executing--;
+		}
+	}
+
+	// Update task and schedule next attempt if needed
+	private async finalize(
+		tx: Db.Tx,
+		task: PartialTask,
+		startedOnTime: Date,
+		status: TaskStatus,
+		errorMessage?: string,
+	): Promise<void> {
+		const attemptCount = task.attempt_count + 1;
+		const body: AnyObject = {
+			started_on__time: startedOnTime,
+			ended_on__time: new Date(),
+			status,
+			attempt_count: attemptCount,
+			...(errorMessage != null && { error_message: errorMessage }),
+		};
+
+		// Re-enqueue if the task failed but has retries left, remember that
+		// attemptCount includes the initial attempt while attempt_limit does not
+		if (status === 'failed' && attemptCount < task.attempt_limit) {
+			body.status = 'queued';
+
+			// Schedule next attempt using exponential backoff
+			body.is_scheduled_to_execute_on__time =
+				this.getNextAttemptTime(attemptCount);
+		}
+
+		// Patch current task
+		await this.client.patch({
+			resource: 'task',
+			passthrough: {
+				tx,
+				req: permissions.root,
+			},
+			id: task.id,
+			body,
+		});
+
+		// Create new task with same configuration if previous
+		// iteration completed and has a cron expression
+		if (
+			['failed', 'succeeded'].includes(body.status) &&
+			task.is_scheduled_with__cron_expression != null
+		) {
+			await this.client.post({
+				resource: 'task',
+				passthrough: {
+					tx,
+					req: permissions.root,
+				},
+				body: {
+					attempt_limit: task.attempt_limit,
+					is_created_by__actor: task.is_created_by__actor,
+					is_executed_by__handler: task.is_executed_by__handler,
+					is_executed_with__parameter_set: task.is_executed_with__parameter_set,
+					is_scheduled_with__cron_expression:
+						task.is_scheduled_with__cron_expression,
+					priority: task.priority,
+				},
+			});
+		}
+	}
+
+	// Calculate next attempt time using exponential backoff
+	private getNextAttemptTime(attempt: number): Date | null {
+		const delay = Math.ceil(Math.exp(Math.min(10, attempt))) * 1000;
```

Comment:
I'm pretty sure this is not doing what you expect as by my calculations this would be setting the minimum delay to a bit over 6 hours (e^10)? I think it probably makes sense to use logic more similar to https://github.com/balena-io-modules/pinejs-client-js/blob/v6.14.6/src/index.ts#L1123 for the next delay time

Reply:
@Page- My mistake here was `* 1000`, not sure how I missed it. I didn't come up with the formula myself, I pulled it from graphile/worker: https://worker.graphile.org/docs/exponential-backoff